src/aircrack-ng.c

```diff
@@ -1032,1096 +1218,661 @@ static int atomic_read(read_buf *rb, int fd, int len, void *buf)
 	return (0);
 }
 
-/* Very similar to check_thread but this one is used when loading
- * files for cracking, when all the parameters are set.
- *
- * TODO: Merge both read_thread and check_thread
- */
-static void read_thread(void *arg)
+static int calculate_wep_keystream(unsigned char *body,
+								   int dlen,
+								   struct AP_info *ap_cur,
+								   unsigned char *h80211)
 {
-	/* we don't care if the buffers allocated here are not freed
-	 * since those threads are only created once, and the memory
-	 * is released to the OS automatically when the program exits.
-	 * there's no point in trying to mute valgrind but breaking
-	 * the multithreaded code at the same time. */
-
-	read_buf rb = {0};
-
-	int fd = -1, n, fmt;
-	unsigned z;
-	int eof_notified = 0;
-	// 	int ret=0;
-
-	unsigned char bssid[6];
-	unsigned char dest[6];
-	unsigned char stmac[6];
-	unsigned char *h80211;
-	unsigned char *p;
+	unsigned char clear[2048];
+	int clearsize, i, j, k;
 	int weight[16];
 
-	struct ivs2_pkthdr ivs2;
-	struct ivs2_filehdr fivs2;
-	struct pcap_pkthdr pkh;
-	struct pcap_file_header pfh;
-	struct AP_info *ap_cur;
-	struct ST_info *st_prv, *st_cur;
+	memset(weight, 0, sizeof(weight));
+	memset(clear, 0, sizeof(clear));
 
-	signal(SIGINT, sighandler);
+	/* calculate keystream */
+	k = known_clear(clear, &clearsize, weight, h80211, (size_t)dlen);
+	if (clearsize < (opt.keylen + 3)) return 0;
 
-	ap_cur = NULL;
+	for (j = 0; j < k; j++)
+	{
+		for (i = 0; i < clearsize; i++) clear[i + (32 * j)] ^= body[4 + i];
+	}
 
-	memset(&pkh, 0, sizeof(struct pcap_pkthdr));
-	memset(&pfh, 0, sizeof(struct pcap_file_header));
+	if (k == 1)
+	{
+		if (ap_cur->ptw_clean == NULL)
+		{
+			ap_cur->ptw_clean = PTW_newattackstate();
```

Comment:
Note these two will (seemingly) use up a lot of memory and previously we only did this in read_thread.
I'm not sure if there is a code path to reach this though in check_mode now.